src/Tasks/Microsoft.NET.Build.Tasks/ResourceUpdater.cs

```diff
@@ -0,0 +1,320 @@
+                }
+            }
+
+            return this;
+        }
+
+        private const ushort LangID_LangNeutral_SublangNeutral = 0;
+
+        private static bool IsIntResource(IntPtr lpType)
+        {
+            return ((uint)lpType >> 16) == 0;
+        }
+
+        /// <summary>
+        /// Add a language-neutral integer resource from a byte[] with
+        /// a particular type and name. This will not modify the
+        /// target until Update() is called.
+        /// Throws an InvalidOperationException if Update() was already called.
+        /// </summary>
+        public ResourceUpdater AddResource(byte[] data, IntPtr lpType, IntPtr lpName)
+        {
+            if (hUpdate == IntPtr.Zero)
+            {
+                ThrowExceptionForInvalidUpdate();
+            }
+
+            if (!IsIntResource(lpType) || !IsIntResource(lpName))
+            {
+                throw new ArgumentException("AddResource can only be used with integer resource types.");
+            }
+
+            if (!UpdateResource(hUpdate, lpType, lpName, LangID_LangNeutral_SublangNeutral, data, (uint)data.Length))
+            {
+                ThrowExceptionForLastWin32Error();
+            }
+
+            return this;
+        }
+
+        /// <summary>
+        /// Write the pending resource updates to the target PE
+        /// file. After this, the ResourceUpdater no longer maintains
+        /// an update handle, and can not be used for further updates.
+        /// Throws an InvalidOperationException if Update() was already called.
+        /// </summary>
+        public void Update()
+        {
+            if (hUpdate == IntPtr.Zero)
+            {
+                ThrowExceptionForInvalidUpdate();
+            }
+
+            try
+            {
+                if (!EndUpdateResource(hUpdate, false))
+                {
+                    ThrowExceptionForLastWin32Error();
+                }
+            }
+            finally
+            {
+                hUpdate = IntPtr.Zero;
+            }
+        }
+
+
+        private bool EnumTypesCallback(IntPtr hModule, IntPtr lpType, IntPtr lParam)
+        {
+            var enumNamesCallback = new EnumResNameProc(EnumNamesCallback);
+            if (!EnumResourceNames(hModule, lpType, enumNamesCallback, lParam))
+            {
+                ThrowExceptionForLastWin32Error();
+            }
+
+            return true;
+        }
+
+        private bool EnumNamesCallback(IntPtr hModule, IntPtr lpType, IntPtr lpName, IntPtr lParam)
+        {
+            var enumLanguagesCallback = new EnumResLangProc(EnumLanguagesCallback);
+            if (!EnumResourceLanguages(hModule, lpType, lpName, enumLanguagesCallback, lParam))
+            {
+                ThrowExceptionForLastWin32Error();
+            }
+
+            return true;
+        }
+
+        private bool EnumLanguagesCallback(IntPtr hModule, IntPtr lpType, IntPtr lpName, ushort wLang, IntPtr lParam)
```

Comment:
I would prefer more descriptive names for the Enum functions. They do enum, but they will eventually copy the resources as well.
Maybe something like:
EnumAndCopyToTarget_LanguagesCallback (NamesCallback, …).
With these names I simply expected it to build a list somewhere which the top caller would go over... which it doesn't.